engines/sci/graphics/gfxdrivers.cpp

```diff
@@ -63,52 +92,277 @@ GfxDefaultDriver::GfxDefaultDriver(uint16 screenWidth, uint16 screenHeight) : Gf
+	assert(dest);
+	assert(size <= (uint32)(_screenW * _screenH));
+
+	// SCI 0 should not make calls to this method (except when using palette mods), but we have to know if it does...
+	if (!_currentBitmap)
+		error("GfxDefaultDriver::copyDataFromCurrentBitmap(): unexpected call");
+
+	// I have changed the implementation a bit from what the engine did before. For non-rgb rendering
+	// it would call OSystem::lockScreen() and then memcpy the data from there (which avoided the need
+	// for the extra bitmap buffer). However, OSystem::lockScreen() is meant more as an update method
+	// for the screen, the call to OSystem::unlockScreen() will turn the whole screen dirty (to be
+	// updated on the next OSysten::updateScreen() call. This is not what we need here, so I rather use
+	// the extra bitmap buffer (which is required for rgb rendering anyway).
+	memcpy(dest, _currentBitmap, size);
 }
 
-SCI0_DOSPreVGADriver::~SCI0_DOSPreVGADriver() {
-	delete[] _compositeBuffer;
+void GfxDefaultDriver::copyCurrentPalette(byte *dest, int start, int num) const {
+	GFXDRV_ASSERT_READY;
+
+	if (_pixelSize == 1) {
+		GfxDriver::copyCurrentPalette(dest, start, num);
+		return;
+	}
+
+	assert(dest);
+	assert(_currentPalette);
+	assert(start + num <= 256);
+	memcpy(dest + start * 3, _currentPalette + start * 3, num * 3);
 }
 
-bool SCI0_DOSPreVGADriver::checkDriver(const char *const *driverNames, int listSize) {
-	Common::String missing;
-	while (listSize-- && *driverNames) {
-		if (Common::File::exists(*driverNames))
-			return true;
-		if (!missing.empty())
-			missing += " or ";
-		missing += "'" + Common::String(*driverNames) + "'";
-		++driverNames;
+template <typename T> void updateRGBPalette(byte *dest, const byte *src, uint start, uint num, Graphics::PixelFormat &f) {
+	T *dst = &reinterpret_cast<T*>(dest)[start];
+	for (uint i = 0; i < num; ++i) {
+		*dst++ = f.RGBToColor(src[0], src[1], src[2]);
+		src += 3;
 	}
-	warning("Driver file %s not found. Starting game in EGA mode", missing.c_str());
-	return false;
+}
+
+void GfxDefaultDriver::updatePalette(const byte *colors, uint start, uint num, bool skipRGBPalette) {
+	memcpy(_currentPalette + start * 3, colors, num * 3);
+	// If palette mods are on we don't need to update the internal palette,
+	// since the colors have to be generated on the fly anyway.
+	if (skipRGBPalette)
+		return;
+	if (_pixelSize == 4)
+		updateRGBPalette<uint32>(_internalPalette, colors, start, num, _format);
+	else if (_pixelSize == 2)
+		updateRGBPalette<uint16>(_internalPalette, colors, start, num, _format);
+	else
+		error("GfxDefaultDriver::updatePalette(): Unsupported pixel size %d", _pixelSize);
+}
+
+template <typename T> void render(byte *dst, const byte *src, int pitch, int w, int h, const byte *pal) {
+	T *d = reinterpret_cast<T*>(dst);
+	const T *p = reinterpret_cast<const T*>(pal);
+	const byte *s = src;
+	pitch -= w;
+
+	while (h--) {
+		for (int i = 0; i < w; ++i)
+			*d++ = p[*s++];
+		s += pitch;
+	}
+}
```

Comment:
I think this is mostly the same as `Graphics::crossBlitMap()` except with a variable size for the map.